src/pages/ReportsUpload/index.tsx

```diff
@@ -0,0 +1,341 @@
+            }
+
+            return event;
+        });
+    }
+
+    componentWillUnmount() {
+        removeEventListener('beforeunload', this.beforeUnloadHandler);
+    }
+
+    /**
+     * Called by "Close modal" or by the "Cancel button"
+     */
+    handleCloseModel = () => {
+        if (!this.props.uploading) {
+            this.props.history.push('/reports');
+        }
+    };
+
+    actionsOnUploadSuccess = () => {
+        const { timeoutToRedirect } = this.state;
+
+        this.startRedirectTimer();
+
+        if (this.state.timeoutToRedirect === 0) {
+            this.props.history.push('/reports');
+        }
+
+        return (
+            <Button type="button" variant={ ButtonVariant.primary } onClick={ this.handleCloseModel }> Closing in { timeoutToRedirect } </Button>
+        );
+    }
+
+    startRedirectTimer = () => {
+        if (this.redirectTimer === null && this.state.timeoutToRedirect > 0) {
+            this.redirectTimer = setInterval(this.redirectCountDown, 1000);
+        }
+    }
+
+    redirectCountDown = () => {
+        let timeoutToRedirect = this.state.timeoutToRedirect - 1;
+        this.setState({
+            timeoutToRedirect
+        });
+
+        if (timeoutToRedirect === 0) {
+            clearInterval(this.redirectTimer);
+        }
+    }
+
+    handleFormSubmit = (values: FormValues) => {
+        this.setState({
+            showForm: false
+        });
+
+        const upload: Upload = {
+            file: this.props.file,
+            reportName: values.reportName,
+            reportDescription: values.reportDescription,
+            yearOverYearGrowthRatePercentage: values.yearOverYearGrowthRatePercentage,
+            percentageOfHypervisorsMigratedOnYear1: values.percentageOfHypervisorsMigratedOnYear1,
+            percentageOfHypervisorsMigratedOnYear2: values.percentageOfHypervisorsMigratedOnYear2,
+            percentageOfHypervisorsMigratedOnYear3: values.percentageOfHypervisorsMigratedOnYear3
+        };
+
+        const config = {
+            headers: {
+                'X-Requested-With': 'XMLHttpRequest'
+            },
+            onUploadProgress: (progressEvent: any) => {
+                const progress: number = Math.round((progressEvent.loaded / progressEvent.total) * 100);
+                this.props.uploadProgress(progress);
+            }
+        };
+
+        this.props.uploadRequest(upload, config);
+    }
+
+    onFileSelected = (files: File[]): void => {
+        this.props.selectUploadFile(files[0]);
+    };
+
+    validateForm = (values: FormValues) => {
+        const validationSchema = this.getValidationSchema(values);
+        try {
+            validationSchema.validateSync(values, { abortEarly: false });
+            return {};
+        } catch (error) {
+            return this.getErrorsFromValidationError(error);
+        }
+    }
+
+    getValidationSchema = (values: FormValues) => {
```

Comment:
I think a lot of these helper methods that don't rely on checking `this.props` or `this.state` would be better placed as imported functions from another file like `helpers.ts` or something like that. They can be unit tested in isolation and they won't clutter up your components.

Reply:
I've moved `getErrorsFromValidationError` and `validateForm` to a different file called `Utilities/formUtils.tsx`. However, I think I should maintain `getValidationSchema` in the same file because it validates the Form and it will be used just from that component.